In [14]:
# Loading libraries:
import pandas as pd
import numpy as np
import statsmodels.api as sm

# Ignoring future warnings
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

## Excersise 1: Perform the first pass regressions and tabulate the summary statistics

In [15]:
# Load Data
data = pd.read_excel('data_market_indes_and_stocks.xlsx',header=1)

stock_excess_returns = data.iloc[:,2:]
market_excess_returns = data['Market Index']

averaged_stock_excess_returns = np.mean(stock_excess_returns)
averaged_market_excess_returns = np.mean(market_excess_returns)

# Creating empty vectors for values
num_stocks = stock_excess_returns.shape[1]
stocks = stock_excess_returns.columns
a           = pd.Series([None]*num_stocks,index=stocks,dtype=float)
a_t_values  = pd.Series([None]*num_stocks,index=stocks,dtype=float)
b           = pd.Series([None]*num_stocks,index=stocks,dtype=float)
b_t_values  = pd.Series([None]*num_stocks,index=stocks,dtype=float)
r_square    = pd.Series([None]*num_stocks,index=stocks,dtype=float)

In [16]:
# First-pass regression
for stock in stocks:
    y = stock_excess_returns[stock]
    x = sm.add_constant(market_excess_returns,prepend=True)

    model = sm.OLS(y,x)
    model = model.fit()

    # print(model.summary())

    a[stock]            = model.params[0]
    a_t_values[stock]   = model.tvalues[0]
    b[stock]            = model.params[1]
    b_t_values[stock]   = model.tvalues[1]
    r_square[stock]     = model.rsquared

In [17]:
# Inserting resutls into a data frame
first_pass_regression = pd.DataFrame()
first_pass_regression['a']          = a
first_pass_regression['a_t_values'] = a_t_values
first_pass_regression['b']          = b
first_pass_regression['b_t_values'] = b_t_values
first_pass_regression['r_square']   = r_square

first_pass_regression

,a,a_t_values,b,b_t_values,r_square
A,8.996946,0.726009,-0.470429,-0.814290,0.062184
B,-0.631102,-0.038668,0.593774,0.780379,0.057403
C,-0.636219,-0.055213,0.416774,0.775836,0.056775
D,-5.050588,-0.413886,1.379241,2.424471,0.370200
E,0.728982,0.053582,0.901305,1.421048,0.168010
F,-4.527576,-0.454789,1.777023,3.828906,0.594494
G,5.936470,0.325840,0.663339,0.780995,0.057489
H,-2.410250,-0.265253,1.911165,4.511642,0.670564
I,5.918236,0.636951,2.082516,4.807727,0.698015


## Excersise 2: Specify the hypotheses for a test of the second-pass regression of the SML
The hypotheses for the second-pass regression for the SML are: <br> 1) The intercept is zero. <br> 2) The slope is equal to the average return on the index portfolio. 

## Excersise 3: Perform the second-pass SML regression by regressing the average excess return for each portfolio on its betas

In [18]:
# averaged_stock_excess_returns = gamma_0 + gamma_1 * b + e

x = sm.add_constant(first_pass_regression['b'],prepend=True)
y = averaged_stock_excess_returns
x.index=range(num_stocks)
y.index=range(num_stocks)
model = sm.OLS(y,x)
model = model.fit()

# print(model.summary())

print('gamma_0 = ' + str(model.params[0]))
print('gamma_1 = ' + str(model.params[1]))
print('sample-average risk premium: rM - rf = ' + str(averaged_market_excess_returns))

gamma_0 = 3.922998719398871
gamma_1 = 5.205334190864971
sample-average risk premium: rM - rf = 8.120833333333332


## Excersise 4: Summirize your test results and compare them to the results reported in the text
The intercept is too high (3.92% per year instead of 0)

The slope is too flat (5.21% instead of a predicted value equal to the sample-average risk premium: rM - rf = 8.12%).

## Excersise 5: Group the nine stocks into three portfolios, maximizing the dispersion of the betas of the three resultant portfolios. Repeat the test, and explain any changes to the results

In [19]:

portfolio_excess_returns = pd.DataFrame()
portfolio_excess_returns['ABC'] = np.mean(stock_excess_returns[['A','B','C']],1)
portfolio_excess_returns['DEG'] = np.mean(stock_excess_returns[['D','E','G']],1)
portfolio_excess_returns['FHI'] = np.mean(stock_excess_returns[['F','H','I']],1)

averaged_portfolio_excess_returns = pd.Series([
    np.mean(portfolio_excess_returns['ABC']),
    np.mean(portfolio_excess_returns['DEG']),
    np.mean(portfolio_excess_returns['FHI']),
],index=[portfolio_excess_returns.columns])

# Creating empty vectors for values
num_portfolios = 3
portfolios = portfolio_excess_returns.columns
a           = pd.Series([None]*num_portfolios,index=portfolios,dtype=float)
a_t_values  = pd.Series([None]*num_portfolios,index=portfolios,dtype=float)
b           = pd.Series([None]*num_portfolios,index=portfolios,dtype=float)
b_t_values  = pd.Series([None]*num_portfolios,index=portfolios,dtype=float)
r_square    = pd.Series([None]*num_portfolios,index=portfolios,dtype=float)

# First-pass regression
for portfolio in portfolios:
    y = portfolio_excess_returns[portfolio]
    x = sm.add_constant(market_excess_returns,prepend=True)

    model = sm.OLS(y,x)
    model = model.fit()

    # print(model.summary())

    a[portfolio]            = model.params[0]
    a_t_values[portfolio]   = model.tvalues[0]
    b[portfolio]            = model.params[1]
    b_t_values[portfolio]   = model.tvalues[1]
    r_square[portfolio]     = model.rsquared

first_pass_regression = pd.DataFrame()
first_pass_regression['a']          = a
first_pass_regression['a_t_values'] = a_t_values
first_pass_regression['b']          = b
first_pass_regression['b_t_values'] = b_t_values
first_pass_regression['r_square']   = r_square

first_pass_regression

,a,a_t_values,b,b_t_values,r_square
ABC,2.576541,0.416032,0.180039,0.623584,0.037430
DEG,0.538288,0.077182,0.981295,3.018145,0.476692
FHI,-0.339864,-0.056226,1.923568,6.826155,0.823310


**Second-pass SML regression** <br> Perform the second-pass SML regression by regressing the average excess return for each portfolio on its betas

In [20]:
# averaged_stock_excess_returns = gamma_0 + gamma_1 * b + e
x = sm.add_constant(first_pass_regression['b'],prepend=True)
y = averaged_portfolio_excess_returns
x.index=range(num_portfolios)
y.index=range(num_portfolios)
model = sm.OLS(y,x)
model = model.fit()

# print(model.summary())

print('gamma_0 = ' + str(model.params[0]))
print('gamma_1 = ' + str(model.params[1]))
print('sample-average risk premium: rM - rf = ' + str(averaged_market_excess_returns))

gamma_0 = 2.622881506147568
gamma_1 = 6.469669670397845
sample-average risk premium: rM - rf = 8.120833333333332


The intercept is too high (2.62% per year instead of 0)

The slope is too flat (6.47% instead of a predicted value equal to the sample-average risk premium: rM - rf = 8.12%). 